grammars/silver/compiler/extension/tuple/README.md

```diff
@@ -0,0 +1,43 @@
+
+Since `tupleElems` occurs on `Type`, we can use the `appType` production to append the individual types (like `Integer`) that may be in the the nested pairs into the `tupleElems` list. It is notable that the `appType` production represents the application of a constructor type (like `Pair`), with constructor `c` and argument `a`. In our new production, we pattern match on the `baseType`of `c` and assign the appropriate value to the `tupleElems` of the top-level `Type`. If the `baseType` of the constructor `c` is a `Pair`, then we know its `argTypes` should have only a single element, the `fst` element in the pair. We then append the `tupleElems` of the argument `a` to the list `argTypes` of `c`, which will either recurse on `a` if it is also the application of a constructor type. If it is not, its `tupleElems` will simply be assigned to the singleton list of the top-level as in the `default production`. 
+
+--
+
+Allie has a few thoughts:
+1. If we use `tupleElems` the way that it is currently implemented, won't there be a `tupleElems` assigned for every single application of a constructor type, even if it is not a pair? This does not seem ideal...
+2. I am not sure if I entirely understand the default production. For tuple purposes, I think it is used to assign a value to `tupleElems` in the case when we have the last element type we need to append?
+
+--
+
+In order to appropriately handle the `typerep` of a tuple, as in productions `tupleTypeExpr` and `tupleExpr` (*Should this also be in `tuplePattern`?*), we assign its value to the direct call of production `tupleType` on the `tupleElems` of the type representation of the `forward`. Please note that we must use `forward` rather than `translation` or `te_translation` in this call because we want to send in a decorated type (*Is this correct?*). Within the `tupleType` production, we properly assign the `typepp` or "pretty print" of the type to be the implosion of the list of types `ts` (that we previously obtained from `tupleElems`). We then forward to the pattern match on `ts`: (1) if we have an empty list of types, we forward to `Unit`; (2) if we have a singleton list, we forward simply to that single type; (3) finally, if we have a proper list of types, we then use the `appType` production to construct the `Pair` type with the head of the list and the recursive call to `tupleType` on the tail of the list.
+
+(*Allie Question: Why do we use/assign attributes `substituted` and `flatrenamed` in `tupleType`?*) 
```

Comment:
This lets substitutions be done without destroying the original structure.  If `substituted` were done by forwarding, the value of `substituted` would have the structure of the pair types to which this forwards rather than the tuple type it originally was.